From b861124ef93753aad171065e7c6b23ac6f41a75f Mon Sep 17 00:00:00 2001
From: "Hung, Merck" <merckhung@gmail.com>
Date: Sun, 16 Mar 2014 02:34:15 +0800
Subject: [PATCH] Hack the DexDump.cpp to list out the location of the MOVE
 instruction.

---
 dexdump/DexDump.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dexdump/DexDump.cpp b/dexdump/DexDump.cpp
index bfb5aa1..18852a6 100644
--- a/dexdump/DexDump.cpp
+++ b/dexdump/DexDump.cpp
@@ -884,6 +884,10 @@ void dumpInstruction(DexFile* pDexFile, const DexCode* pCode, int insnIdx,
             printf("|%04x: nop // spacer", insnIdx);
         }
     } else {
+        if( !strcmp( dexGetOpcodeName(pDecInsn->opcode), "move" ) ) {
+
+					fprintf( stderr, "%06x\n", ((u1*)insns - pDexFile->baseAddr) + insnIdx*2);
+        }
         printf("|%04x: %s", insnIdx, dexGetOpcodeName(pDecInsn->opcode));
     }
 
-- 
1.8.3.2

